local-postgresql-copy/run-pg-restore.sh

```diff
@@ -0,0 +1,210 @@
+
+if [ -z "$spanner_restore_database" ]; then spanner_restore_database="$SPANNER_DATABASE"; fi;
+if [ -z "$schema_file" ]; then schema_file="./backup/schema.sql"; fi;
+if [ -z "$data_dir" ]; then data_dir="./backup/data"; fi;
+
+echo "RESTORING CLOUD SPANNER DATABASE"
+echo "Restoring to Cloud Spanner database $spanner_restore_database. This database must be empty."
+echo "Restoring schema file $schema_file."
+echo "Restoring data from $data_dir."
+echo "--- WARNING: pg_restore is executed with the options --clean --create. ---"
+echo "--- ANY EXISTING DATA IN THE PostgreSQL DATABASE WILL BE LOST! ---"
+
+echo
+read -p "Continue? [Yn]" -n 1 -r
+echo
+if [[ $REPLY =~ [Nn]$ ]]
+then
+    [[ "$0" = "$BASH_SOURCE" ]] && exit 1 || return 1
+fi
+
+# Check whether the Cloud Spanner database is empty. That is; no user-tables.
+table_count=$(psql -v ON_ERROR_STOP=1 -h "$PGADAPTER_HOST" -p "$PGADAPTER_PORT" \
+  -d "$spanner_restore_database" \
+  -qAtX \
+  -c "select count(1) from information_schema.tables where table_schema in (select schema_name from information_schema.schemata where not schema_owner='spanner_system')")
+if [[ ! $table_count = "0" ]]
+then
+  echo "Number of tables found in the restore database: $table_count"
+  echo "The destination Cloud Spanner database is not empty."
+  echo "This script only supports restoring into an empty database."
+  [[ "$0" = "$BASH_SOURCE" ]] && exit 1 || return 1
+fi
+
+# Restore the database. --clean and --create means 'drop-and-recreate' if the database exists.
+# We 'connect' to the postgres database here, because we need to connect to a database in order
+# to drop another database.
+# Note that we do a `|| true` to ensure that the script continues even if pg_restore returns
+# a non-zero exit code. That is because pg_restore will often return some warnings, and that
+# again will cause it to return a non-zero exit code.
+PGPASSWORD="$POSTGRES_PASSWORD" pg_restore -v \
+  -U "$POSTGRES_USER" \
+  -h "$POSTGRES_HOST" \
+  -p "$POSTGRES_PORT" \
+  -d postgres \
+  --create --clean \
+  --format=directory "$data_dir" || true
+
+# Create the schema on the Cloud Spanner database. We run the script between
+# START BATCH DDL and RUN BATCH to run the entire script as one batch.
+echo "Creating schema on Cloud Spanner database $spanner_restore_database"
+psql -v ON_ERROR_STOP=1 -a \
+  --host "$PGADAPTER_HOST" \
+  --port "$PGADAPTER_PORT" \
+  -d "$spanner_restore_database" \
+  -c "start batch ddl" \
+  -f "$schema_file" \
+  -c "run batch"
+
+# Re-import the Cloud Spanner database as foreign schemas in the PostgreSQL database.
+# This will allow us to query the Cloud Spanner information_schema using the PostgreSQL query engine,
+# which again allows us to execute a recursive query on the information_schema.tables table in Cloud Spanner.
+echo "Importing the Cloud Spanner schema as a foreign schema in the PostgreSQL database"
+PGPASSWORD="$POSTGRES_PASSWORD" psql -v ON_ERROR_STOP=1 \
```

Comment:
Minor readability nit -- it looks like psql is only invoked in a few (2?) ways?  Would it be more readable to create well-named variables containing the environment/args for each, rather than repeating all of the args each time?

Reply:
I'm going to defer this to a later pull request in order to be able to finish both this and the Spring Data JPA sample before my vacation :-)